File: funtime/pylib/match3_draw.py
"""match3_draw.py - everything you can see in Match Three, in Python.

Six different SHAPES, not six colours - a circle, a square, a triangle, a
diamond, a star and a cross. That way the game works in black and white, and
it works for a colour-blind player too.
"""

import math

from match3_rules import GRID_SIZE, EMPTY, gem_index, is_inside_board

CELL = 34
BOARD_MARGIN = 8

COLOR_INK = "#111111"
COLOR_PAPER = "#ffffff"
COLOR_FAINT = "#e2e2e2"


def board_pixel_size():
    """How big the square canvas must be."""
    return BOARD_MARGIN * 2 + GRID_SIZE * CELL


def cell_left(column):
    """The left edge of one square."""
    return BOARD_MARGIN + column * CELL


def cell_top(row):
    """The top edge of one square."""
    return BOARD_MARGIN + row * CELL


def cell_at_pixel(x, y):
    """Which square did the player click? (column, row), or None."""
    column = int((x - BOARD_MARGIN) // CELL)
    row = int((y - BOARD_MARGIN) // CELL)
    if not is_inside_board(column, row):
        return None
    return (column, row)


def clear_canvas(ctx, width, height, color):
    """Paint the whole canvas one flat colour."""
    ctx.fillStyle = color
    ctx.fillRect(0, 0, width, height)


def draw_shape(ctx, shape, x, y, size):
    """One of the six shapes, drawn in the middle of a square.

    ALGORITHM: one branch per shape. Shapes 0, 2 and 4 are solid and 1, 3 and
    5 are outlines, so even two similar shapes never look alike.
    """
    r = size / 2
    ctx.strokeStyle = COLOR_INK
    ctx.fillStyle = COLOR_INK
    ctx.lineWidth = 2.5

    if shape == 0:                      # solid circle
        ctx.beginPath()
        ctx.arc(x, y, r, 0, math.pi * 2)
        ctx.fill()

    elif shape == 1:                    # hollow square
        ctx.strokeRect(x - r, y - r, r * 2, r * 2)

    elif shape == 2:                    # solid triangle
        ctx.beginPath()
        ctx.moveTo(x, y - r)
        ctx.lineTo(x + r, y + r)
        ctx.lineTo(x - r, y + r)
        ctx.closePath()
        ctx.fill()

    elif shape == 3:                    # hollow diamond
        ctx.beginPath()
        ctx.moveTo(x, y - r)
        ctx.lineTo(x + r, y)
        ctx.lineTo(x, y + r)
        ctx.lineTo(x - r, y)
        ctx.closePath()
        ctx.stroke()

    elif shape == 4:                    # solid star
        ctx.beginPath()
        for point in range(10):
            reach = r if point % 2 == 0 else r * 0.45
            angle = -math.pi / 2 + point * math.pi / 5
            px = x + math.cos(angle) * reach
            py = y + math.sin(angle) * reach
            if point == 0:
                ctx.moveTo(px, py)
            else:
                ctx.lineTo(px, py)
        ctx.closePath()
        ctx.fill()

    else:                               # hollow cross
        ctx.beginPath()
        ctx.moveTo(x - r, y - r)
        ctx.lineTo(x + r, y + r)
        ctx.moveTo(x + r, y - r)
        ctx.lineTo(x - r, y + r)
        ctx.stroke()


def draw_grid(ctx):
    """The faint lines between the squares."""
    ctx.strokeStyle = COLOR_FAINT
    ctx.lineWidth = 1
    for i in range(GRID_SIZE + 1):
        ctx.beginPath()
        ctx.moveTo(BOARD_MARGIN + i * CELL, BOARD_MARGIN)
        ctx.lineTo(BOARD_MARGIN + i * CELL, BOARD_MARGIN + GRID_SIZE * CELL)
        ctx.moveTo(BOARD_MARGIN, BOARD_MARGIN + i * CELL)
        ctx.lineTo(BOARD_MARGIN + GRID_SIZE * CELL, BOARD_MARGIN + i * CELL)
        ctx.stroke()


def draw_picked(ctx, square):
    """The heavy box round the shape waiting to be swapped."""
    ctx.strokeStyle = COLOR_INK
    ctx.lineWidth = 3
    ctx.strokeRect(cell_left(square["column"]) + 2, cell_top(square["row"]) + 2,
                   CELL - 4, CELL - 4)


def draw_cursor(ctx, cursor):
    """The dashed box showing where the keyboard is pointing."""
    ctx.strokeStyle = COLOR_INK
    ctx.lineWidth = 2
    ctx.setLineDash([4, 3])
    ctx.strokeRect(cell_left(cursor["column"]) + 4, cell_top(cursor["row"]) + 4,
                   CELL - 8, CELL - 8)
    ctx.setLineDash([])


def draw_message(ctx, width, height, title, subtitle):
    """Big centred words across the board."""
    ctx.fillStyle = "rgba(255, 255, 255, 0.93)"
    ctx.fillRect(0, height / 2 - 40, width, 80)
    ctx.fillStyle = COLOR_INK
    ctx.textAlign = "center"

    scale = width / board_pixel_size()
    title_size = max(14, round(26 * scale))
    subtitle_size = max(9, round(12 * scale))

    ctx.font = "bold %dpx monospace" % title_size
    ctx.fillText(title, width / 2, height / 2 - 2)
    ctx.font = "%dpx monospace" % subtitle_size
    ctx.fillText(subtitle, width / 2, height / 2 + title_size * 0.8)
    ctx.textAlign = "left"


def render_game(ctx, state):
    """Draw one complete frame."""
    size = board_pixel_size()
    clear_canvas(ctx, size, size, COLOR_PAPER)
    draw_grid(ctx)

    for row in range(GRID_SIZE):
        for column in range(GRID_SIZE):
            shape = state["board"][gem_index(column, row)]
            if shape != EMPTY:
                draw_shape(ctx, shape, cell_left(column) + CELL / 2,
                           cell_top(row) + CELL / 2, CELL - 16)

    if state["picked"]:
        draw_picked(ctx, state["picked"])
    if not state["is_over"]:
        draw_cursor(ctx, state["cursor"])

    ctx.strokeStyle = COLOR_INK
    ctx.lineWidth = 2
    ctx.strokeRect(BOARD_MARGIN - 1, BOARD_MARGIN - 1,
                   GRID_SIZE * CELL + 2, GRID_SIZE * CELL + 2)

    if state["is_paused"]:
        draw_message(ctx, size, size, "PAUSED", "Press P to carry on")

File: funtime/lib/match3-draw.js
/* ============================================================
   match3-draw.js — everything you can see in Match Three

   Six different SHAPES, not six colours — a circle, a square, a triangle, a
   diamond, a star and a cross. That way the game works in black and white,
   and it works for a colour-blind player too, which is worth remembering
   whenever you are tempted to make colour carry the meaning.
   ============================================================ */

const CELL = 34;
const BOARD_MARGIN = 8;

const COLOR_INK = '#111111';
const COLOR_PAPER = '#ffffff';
const COLOR_FAINT = '#e2e2e2';

/** boardPixelSize — how big the square canvas must be. */
function boardPixelSize() {
    return BOARD_MARGIN * 2 + GRID_SIZE * CELL;
}

/** cellLeft / cellTop — the top-left corner of one square. */
function cellLeft(column) { return BOARD_MARGIN + column * CELL; }
function cellTop(row) { return BOARD_MARGIN + row * CELL; }

/** cellAtPixel — which square did the player click? null if none. */
function cellAtPixel(x, y) {
    const column = Math.floor((x - BOARD_MARGIN) / CELL);
    const row = Math.floor((y - BOARD_MARGIN) / CELL);
    if (!isInsideBoard(column, row)) {
        return null;
    }
    return { column: column, row: row };
}

/** clearCanvas — paint the whole canvas one flat colour. */
function clearCanvas(ctx, width, height, color) {
    ctx.fillStyle = color;
    ctx.fillRect(0, 0, width, height);
}

/**
 * drawShape — one of the six shapes, drawn in the middle of a square.
 * INPUT: ctx, shape (0 to 5), x and y of the middle, size.
 * ALGORITHM: one branch per shape. Shapes 0, 2 and 4 are solid and 1, 3 and 5
 *            are outlines, so even two similar shapes never look alike.
 */
function drawShape(ctx, shape, x, y, size) {
    const r = size / 2;
    ctx.strokeStyle = COLOR_INK;
    ctx.fillStyle = COLOR_INK;
    ctx.lineWidth = 2.5;

    if (shape === 0) {                       /* solid circle */
        ctx.beginPath();
        ctx.arc(x, y, r, 0, Math.PI * 2);
        ctx.fill();

    } else if (shape === 1) {                /* hollow square */
        ctx.strokeRect(x - r, y - r, r * 2, r * 2);

    } else if (shape === 2) {                /* solid triangle */
        ctx.beginPath();
        ctx.moveTo(x, y - r);
        ctx.lineTo(x + r, y + r);
        ctx.lineTo(x - r, y + r);
        ctx.closePath();
        ctx.fill();

    } else if (shape === 3) {                /* hollow diamond */
        ctx.beginPath();
        ctx.moveTo(x, y - r);
        ctx.lineTo(x + r, y);
        ctx.lineTo(x, y + r);
        ctx.lineTo(x - r, y);
        ctx.closePath();
        ctx.stroke();

    } else if (shape === 4) {                /* solid star */
        ctx.beginPath();
        for (let point = 0; point < 10; point++) {
            const reach = point % 2 === 0 ? r : r * 0.45;
            const angle = -Math.PI / 2 + point * Math.PI / 5;
            const px = x + Math.cos(angle) * reach;
            const py = y + Math.sin(angle) * reach;
            if (point === 0) { ctx.moveTo(px, py); } else { ctx.lineTo(px, py); }
        }
        ctx.closePath();
        ctx.fill();

    } else {                                 /* hollow cross */
        ctx.beginPath();
        ctx.moveTo(x - r, y - r);
        ctx.lineTo(x + r, y + r);
        ctx.moveTo(x + r, y - r);
        ctx.lineTo(x - r, y + r);
        ctx.stroke();
    }
}

/** drawGrid — the faint lines between the squares. */
function drawGrid(ctx) {
    ctx.strokeStyle = COLOR_FAINT;
    ctx.lineWidth = 1;
    for (let i = 0; i <= GRID_SIZE; i++) {
        ctx.beginPath();
        ctx.moveTo(BOARD_MARGIN + i * CELL, BOARD_MARGIN);
        ctx.lineTo(BOARD_MARGIN + i * CELL, BOARD_MARGIN + GRID_SIZE * CELL);
        ctx.moveTo(BOARD_MARGIN, BOARD_MARGIN + i * CELL);
        ctx.lineTo(BOARD_MARGIN + GRID_SIZE * CELL, BOARD_MARGIN + i * CELL);
        ctx.stroke();
    }
}

/** drawPicked — the heavy box round the shape waiting to be swapped. */
function drawPicked(ctx, square) {
    ctx.strokeStyle = COLOR_INK;
    ctx.lineWidth = 3;
    ctx.strokeRect(cellLeft(square.column) + 2, cellTop(square.row) + 2, CELL - 4, CELL - 4);
}

/** drawCursor — the dashed box showing where the keyboard is pointing. */
function drawCursor(ctx, cursor) {
    ctx.strokeStyle = COLOR_INK;
    ctx.lineWidth = 2;
    ctx.setLineDash([4, 3]);
    ctx.strokeRect(cellLeft(cursor.column) + 4, cellTop(cursor.row) + 4, CELL - 8, CELL - 8);
    ctx.setLineDash([]);
}

/** drawMessage — big centred words across the board. */
function drawMessage(ctx, width, height, title, subtitle) {
    ctx.fillStyle = 'rgba(255, 255, 255, 0.93)';
    ctx.fillRect(0, height / 2 - 40, width, 80);
    ctx.fillStyle = COLOR_INK;
    ctx.textAlign = 'center';

    const scale = width / boardPixelSize();
    const titleSize = Math.max(14, Math.round(26 * scale));
    const subtitleSize = Math.max(9, Math.round(12 * scale));

    ctx.font = 'bold ' + titleSize + 'px monospace';
    ctx.fillText(title, width / 2, height / 2 - 2);
    ctx.font = subtitleSize + 'px monospace';
    ctx.fillText(subtitle, width / 2, height / 2 + titleSize * 0.8);
    ctx.textAlign = 'left';
}

/** renderGame — draw one complete frame. */
function renderGame(ctx, state) {
    const size = boardPixelSize();
    clearCanvas(ctx, size, size, COLOR_PAPER);
    drawGrid(ctx);

    for (let row = 0; row < GRID_SIZE; row++) {
        for (let column = 0; column < GRID_SIZE; column++) {
            const shape = state.board[gemIndex(column, row)];
            if (shape !== EMPTY) {
                drawShape(ctx, shape, cellLeft(column) + CELL / 2,
                          cellTop(row) + CELL / 2, CELL - 16);
            }
        }
    }

    if (state.picked) {
        drawPicked(ctx, state.picked);
    }
    if (!state.isOver) {
        drawCursor(ctx, state.cursor);
    }

    ctx.strokeStyle = COLOR_INK;
    ctx.lineWidth = 2;
    ctx.strokeRect(BOARD_MARGIN - 1, BOARD_MARGIN - 1,
                   GRID_SIZE * CELL + 2, GRID_SIZE * CELL + 2);

    if (state.isPaused) {
        drawMessage(ctx, size, size, 'PAUSED', 'Press P to carry on');
    }
}
